api/src/main/java/uk/gov/hmcts/payment/api/controllers/OrderController.java

```diff
@@ -0,0 +1,254 @@
+        Function<String, Optional<IdempotencyKeys>> getIdempotencyKey = idempotencyKeyToCheck -> idempotencyKeysRepository.findByIdempotencyKey(idempotencyKeyToCheck);
+
+        Function<IdempotencyKeys, ResponseEntity<?>> validateHashcodeForRequest = idempotencyKeys -> {
+
+            OrderPaymentBo responseBO;
+            try {
+                if (!idempotencyKeys.getRequest_hashcode().equals(orderPaymentDto.hashCodeWithOrderReference(orderReference))) {
+                    return new ResponseEntity<>("Payment already present for idempotency key with different payment details", HttpStatus.CONFLICT); // 409 if hashcode not matched
+                }
+                if (idempotencyKeys.getResponseCode() >= 500) {
+                    return new ResponseEntity<>(idempotencyKeys.getResponseBody(), HttpStatus.valueOf(idempotencyKeys.getResponseCode()));
+                }
+                responseBO = objectMapper.readValue(idempotencyKeys.getResponseBody(), OrderPaymentBo.class);
+            } catch (JsonProcessingException e) {
+                return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
+            }
+            return new ResponseEntity<>(responseBO, HttpStatus.valueOf(idempotencyKeys.getResponseCode())); // if hashcode matched
+        };
+
+        //Idempotency Check
+        Optional<IdempotencyKeys> idempotencyKeysRow = getIdempotencyKey.apply(idempotencyKey);
+        if (idempotencyKeysRow.isPresent()) {
+            ResponseEntity responseEntity = validateHashcodeForRequest.apply(idempotencyKeysRow.get());
+            return responseEntity;
+        }
+
+        //business validations for order
+        PaymentFeeLink order = businessValidationForOrders(orderDomainService.find(orderReference), orderPaymentDto);
+
+        //PBA Payment
+        OrderPaymentBo orderPaymentBo = null;
+        ResponseEntity responseEntity;
+        String responseJson;
+        try {
+            orderPaymentBo = orderDomainService.addPayments(order, orderPaymentDto);
+            HttpStatus httpStatus = orderPaymentBo.getStatus().equalsIgnoreCase(FAILED) ? HttpStatus.PAYMENT_REQUIRED : HttpStatus.CREATED; //402 for failed Payment scenarios
+            responseEntity = new ResponseEntity<>(orderPaymentBo, httpStatus);
+            responseJson = objectMapper.writeValueAsString(orderPaymentBo);
+        } catch (LiberataServiceTimeoutException liberataServiceTimeoutException) {
+            responseEntity = new ResponseEntity<>(liberataServiceTimeoutException.getMessage(), HttpStatus.GATEWAY_TIMEOUT);
+            responseJson = liberataServiceTimeoutException.getMessage();
+        }
+
+        //Create Idempotency Record
+        return createIdempotencyRecord(objectMapper, idempotencyKey, orderReference, responseJson, responseEntity, orderPaymentDto);
+    }
+
+    private PaymentFeeLink businessValidationForOrders(PaymentFeeLink order, OrderPaymentDto orderPaymentDto) {
```

Comment:
can't we move this method into service layer? Ideally, Controller classes should only have endpoints in them

Reply:
Moved